### PR TITLE
[Core] Adjoint response function base class

### PR DESCRIPTION
**Pull request by msandre**

### CODE REVIEW

File: kratos/solving_strategies/response_functions/adjoint_response_function.h [hunk 1]
```diff
@@ -0,0 +1,286 @@
...
+     *                                   element's residual w.r.t. primal.
+     * @param[out]    rResponseGradient  the gradient of the response function.
+     * @param[in]     rProcessInfo       the current process info.
+     */
+    virtual void CalculateGradient(const Element& rAdjointElement,
+                                   const Matrix& rResidualGradient,
+                                   Vector& rResponseGradient,
+                                   ProcessInfo& rProcessInfo)
+    {
+        KRATOS_ERROR << "Calling base class response function method.\n";
```
Comment (armingeiser) [inline kratos/solving_strategies/response_functions/adjoint_response_function.h]: Why do you distinguish between these functions with this error message and the pure virtual function ```CalculateValue```?
From a developer point of you you are forced to implement both, but for one case you notice at run time, for the other case at compile time.

File: kratos/solving_strategies/response_functions/adjoint_response_function.h [hunk 2]
```diff
@@ -0,0 +1,286 @@
...
+    /// Calculate the local gradient w.r.t. primal solution.
+    /**
+     * @param[in]     rAdjointElement    the adjoint element.
+     * @param[in]     rResidualGradient  the transposed gradient of the
+     *                                   element's residual w.r.t. primal.
+     * @param[out]    rResponseGradient  the gradient of the response function.
+     * @param[in]     rProcessInfo       the current process info.
+     */
+    virtual void CalculateGradient(const Element& rAdjointElement,
+                                   const Matrix& rResidualGradient,
```
Comment (armingeiser) [inline kratos/solving_strategies/response_functions/adjoint_response_function.h]: rElementResidualGradient

File: kratos/solving_strategies/response_functions/adjoint_response_function.h [hunk 3]
```diff
@@ -0,0 +1,286 @@
...
+    /// Calculate the local gradient w.r.t. first derivatives of primal solution.
+    /**
+     * @param[in]     rAdjointElement    the adjoint element.
+     * @param[in]     rResidualGradient  the transposed gradient of the
+     *                                   element's residual w.r.t. first derivatives.
+     * @param[out]    rResponseGradient  the gradient of the response function.
+     * @param[in]     rProcessInfo       the current process info.
+     */
+    virtual void CalculateFirstDerivativesGradient(const Element& rAdjointElement,
+                                                   const Matrix& rResidualGradient,
```
Comment (armingeiser) [inline kratos/solving_strategies/response_functions/adjoint_response_function.h]: rElementResidualGradient

File: kratos/solving_strategies/response_functions/adjoint_response_function.h [hunk 4]
```diff
@@ -0,0 +1,286 @@
...
+    /// Calculate the local gradient w.r.t. first derivatives of primal solution.
+    /**
+     * @param[in]     rAdjointCondition  the adjoint condition.
+     * @param[in]     rResidualGradient  the transposed gradient of the
+     *                                   condition's residual w.r.t. first derivatives.
+     * @param[out]    rResponseGradient  the gradient of the response function.
+     * @param[in]     rProcessInfo       the current process info.
+     */
+    virtual void CalculateFirstDerivativesGradient(const Condition& rAdjointCondition,
+                                                   const Matrix& rResidualGradient,
```
Comment (armingeiser) [inline kratos/solving_strategies/response_functions/adjoint_response_function.h]: rConditionResidualGradient

**Comment by armingeiser** [comment]
UpdateSensitivities() is not there, so I assume you want to do these things in the FinalizeSolutionStep?

Also there where some helper functions to assemble the sensitivity contribution, they could be useful to have in the base.

File: kratos/solving_strategies/response_functions/adjoint_response_function.h (hunk 1 above)
Comment (msandre) [reply kratos/solving_strategies/response_functions/adjoint_response_function.h]: Yes, that is the only distinction. With pure virtual we are making `CalculateValue` mandatory. The other methods are optional, provided you don't try to use one of them. For example, you might use `CalculateGradient` for a static response, but not `CalculateSecondDerivativesGradient`. As long as you don't try passing your static response to a client which expects second derivatives gradient you will be fine.

**Comment by msandre** [comment]
I cleaned up the variables. The `UpdateSensitivities` has be clarified in an internal discussion. For now the plan is to omit it from this base class.

**Comment by armingeiser** [comment]
Good from my side.

**Comment by msandre** [comment]
Based on the feedback of the @KratosMultiphysics/technical-committee the response function has been moved to kratos/response_functions. I think this is ready to merge. Can someone approve?

**Comment by jcotela** [comment]
@msandre looks like the appveyor fail is relevant here... Some kind of linking error, I guess related to the EXPORT_API things?

**Comment by msandre** [comment]
Removing the export api fixed it. Thanks @jcotela.

**Review by jcotela — CHANGES_REQUESTED** [review]
I agree with the general design, but I have one specific concern: I've found that some functions have parameters marked as [in] but passed as non-const reference... I'd try to avoid that if possible (I know sometimes it is not).

In particular, I strongly recommend enforcing const-ness for ProcessInfo& arguments in the interface. ProcessInfo not being const has been a consistent source of race conditions for elements and schemes (when using it within a parallel loop) and I'd like to avoid that here if possible.

**Comment by msandre** [comment]
@jcotela done.

**Review by jcotela — APPROVED** [review]

**Pull request merged**